### PR TITLE
[Update] 1/N for v0.15.1 Implement and register Fused MoE Kunlun kernels using OOT method

### PR DESCRIPTION
**Pull request by xyDong0223**
### Summary

This PR refactors the Kunlun FusedMoE operator registration from **Monkey Patching** to vLLM's official **Out-of-Tree (OOT) CustomOp registration mechanism**, making the codebase more standardized, maintainable, and aligned with the vLLM v0.15.1 plugin architecture.

### Motivation

In the previous implementation, Kunlun's FusedMoE operators were injected by directly monkey-patching `UnquantizedFusedMoEMethod` and `FusedMoE` in `vllm.model_executor.layers.fused_moe.layer`. This approach had several drawbacks:

- **Highly invasive**: Directly mutating global references inside vLLM internals, making it fragile across version upgrades
- **Code redundancy**: Required duplicating the entire `FusedMoE.__init__` parameter list in `KunlunFusedMoE`, increasing maintenance cost
- **Non-compliant with plugin standards**: Did not leverage vLLM's `CustomOp.register_oot` decorator or `entry_points` discovery mechanism

#### 1. Adopt `CustomOp.register_oot` decorator for operator registration
```python
@CustomOp.register_oot(name="UnquantizedFusedMoEMethod")
class KunlunUnquantizedFusedMoEMethod(UnquantizedFusedMoEMethod):
    ...
```
With the OOT decorator, vLLM automatically replaces `UnquantizedFusedMoEMethod` with `KunlunUnquantizedFusedMoEMethod` at load time — no manual monkey patching required.

#### 2. Enable Monolithic Mode for forward inference
```python
@property
def is_monolithic(self) -> bool:
    return True
```
- Setting `is_monolithic = True` causes `FusedMoE.forward_impl` to call `apply_monolithic(layer, x, router_logits)` directly
- `router_logits` are passed straight to `KunlunOps.fused_moe`, allowing the Kunlun device kernels to handle routing on-chip, avoiding unnecessary host-device data transfers

#### 3. Skip Triton kernel initialization
```python
def process_weights_after_loading(self, layer):
    FusedMoEMethodBase.process_weights_after_loading(self, layer)
```
Kunlun does not rely on Triton kernels, so we call the base class method directly, bypassing `_setup_kernel()`.

#### 4. Remove redundant `KunlunFusedMoE` subclass
Deleted ~**100 lines** of the `KunlunFusedMoE(FusedMoE)` class, which previously existed solely to force-set `KunlunUnquantizedFusedMoEMethod` in `__init__`. This is now handled entirely by the OOT registration mechanism.

#### 5. Register plugin via `entry_points`
Added a new `vllm.plugins` entry point in `setup.py`:
```python
"vllm.plugins": [
    "kunlun_fused_moe = vllm_kunlun.ops.fused_moe:register_kunlun_fused_moe_ops"
],
```
vLLM will automatically discover and load the Kunlun FusedMoE operators at startup.

#### 6. Code style cleanup
- Unified string quoting to double quotes in `setup.py`
- Normalized import ordering and blank line formatting

### Files Changed

| File | Description |
|------|-------------|
| `setup.py` | Added `vllm.plugins` entry point; unified code style |
| `vllm_kunlun/ops/fused_moe/__init__.py` | Added OOT registration entry function `register_kunlun_fused_moe_ops` |
| `vllm_kunlun/ops/fused_moe/layer.py` | Refactored to OOT + Monolithic mode; removed monkey patch and redundant class |

### Impact

- ✅ **No functional change**: The actual FusedMoE computation logic (`KunlunOps.fused_moe`) remains unchanged
- ✅ **Improved compatibility**: Follows vLLM's OOT plugin conventions, reducing adaptation cost for future version upgrades
- ✅ **Cleaner codebase**: Net reduction of ~100+ lines by eliminating the redundant `KunlunFusedMoE` class and monkey patch logic
- ✅ **EP support preserved**: Both Expert Parallelism (EP) and non-EP paths are fully retained

### Note

This is **1/N** of the v0.15.1 update series (4/N branch). Subsequent PRs will continue to align the remaining Kunlun plugin components with the vLLM v0.15.1 OOT architecture.

---

## Checklist (Required)

Before submitting this PR, please ensure that all the following items are completed:

- [x] All code changes pass the [`pre-commit`](https://github.com/baidu/vLLM-Kunlun/blob/main/CONTRIBUTING.md) checks.
- [x] Commits are signed off using `git commit -s`.
- [x] The PR title is properly classified (see below).

---

## PR Type

Please prefix the PR title with one or more of the following labels to help reviewers quickly understand the nature of the change:

- `[Feature]` – New features or enhancements (e.g. Attention, Communicator, Kernel, Worker, etc.)
- `[Bugfix]` – Bug fixes
- `[CI/Build]` – CI, build system, or infrastructure improvements
- `[Doc]` – Documentation updates or fixes
- `[Misc]` – Other changes that do not fit the above categories (use sparingly)

> **Note:** If the PR spans multiple categories, include all relevant prefixes.

---

<details>
<summary><b>Detailed Checklist (Click to Expand)</b></summary>

<p>Thank you for contributing to <b>vLLM Kunlun</b>!  
To help us maintain high code quality and streamline the review process, please ensure your PR meets the following requirements.</p>

<h3>1. Code Quality</h3>

<ul>
    <li>All linting and formatting checks pass (<code>pre-commit</code>).</li>
    <li>The code is well-structured and sufficiently documented.</li>
    <li>The change is designed with maintainability and readability in mind.</li>
</ul>

<h3>2. Testing</h3>

<ul>
    <li>Relevant unit tests are added or updated.</li>
    <li>Integration tests are included when applicable.</li>
    <li>Existing tests continue to pass.</li>
</ul>

<h3>3. DCO Compliance</h3>

<p>This project follows the
<a href="https://github.com/vllm-project/vllm/blob/main/DCO">Developer Certificate of Origin (DCO)</a>.</p>

<ul>
    <li>All commits include a <code>Signed-off-by:</code> line.</li>
    <li>Use <code>git commit -s</code> to automatically add the sign-off.</li>
</ul>

<h3>4. Review Expectations</h3>

<p>During the review process, maintainers may:</p>

<ul>
    <li>Request code refactoring or additional tests.</li>
    <li>Ask for clarifications on design decisions.</li>
    <li>Suggest performance, stability, or maintainability improvements.</li>
</ul>

<p>We appreciate your patience and collaboration throughout the review process!</p>

</details>

### CODE REVIEW

File: vllm_kunlun/ops/fused_moe/layer.py [hunk 1]
```diff
@@ -1,64 +1,52 @@
...
-#
-# Unless required by applicable law or agreed to in writing, software
-# distributed under the License is distributed on an "AS IS" BASIS,
-# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-# See the License for the specific language governing permissions and
-# limitations under the License.
+"""
+Kunlun optimized FusedMoE - replaces UnquantizedFusedMoEMethod
+Uses monolithic mode to receive router_logits directly and call KunlunOps.fused_moe
+"""
```
[the remark this answers is not in the document]
Comment (xyDong0223) [reply vllm_kunlun/ops/fused_moe/layer.py:4]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/baidu/vLLM-Kunlun/pull/205#discussion_r2807046317)

**Comment by xyDong0223** [comment]
@liwei109 Quant Part might be suitable for this OOT method
